### PR TITLE
Optimize _count type API requests

### PR DESCRIPTION
**Pull request by benwtrent**
When calling `_count` or simply `_search?size=0&track_total_hits=true` with no aggregations, we end up doing too much work on the coordinator, allocating large arrays and keeping unnecessary references to various objects.

This commit optimizes this scenario by only collecting the count and allowing result references to be collected.

### CODE REVIEW

**Comment by benwtrent** [comment]
Lol, CI is green. I am thinking I need to write some tests.

File: docs/changelog/102888.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+pr: 102888
+summary: "{DRAFT} Optimize `_count` type API requests"
```
Comment (benwtrent) [inline docs/changelog/102888.yaml:2]:
```suggestion
summary: "Optimize `_count` type API requests"
```

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java [hunk 2]
```diff
@@ -556,4 +563,103 @@ public synchronized void cancel() {
             consumeListener();
         }
     }
+
+    /**
+     * Optimized phase result consumer that only counts the number of hits and does not
+     * store any other information.
+     */
+    static class CountOnlyQueryPhaseResultConsumer extends SearchPhaseResults<SearchPhaseResult> {
```
Comment (jdconrad) [inline server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:571]: This is complex enough I would have it be its own top-level class.

**Review by jdconrad — APPROVED** [review]
LGTM! I left a few minor comments.

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 3]
```diff
@@ -792,6 +792,17 @@ QueryPhaseResultConsumer newSearchPhaseResults(
         int numShards,
         Consumer<Exception> onPartialMergeFailure
     ) {
+        final int size = request.source() == null || request.source().size() == -1 ? SearchService.DEFAULT_SIZE : request.source().size();
+        // Use the new CountOnlyQueryPhaseResultConsumer for requests without aggs, suggest, etc. things only wanting a total count and
+        // returning no hits
+        if (size == 0
+            && request.source().aggregations() == null
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:799]: @jdconrad there isn't a way for `size==0` && `request.source() == null`. Our default for `size=20`, so for it to be `0` its part of the search source.

File: server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java [hunk 4]
```diff
@@ -43,7 +43,7 @@ class SearchQueryThenFetchAsyncAction extends AbstractSearchAsyncAction<SearchPh
         final Map<String, AliasFilter> aliasFilter,
         final Map<String, Float> concreteIndexBoosts,
         final Executor executor,
-        final QueryPhaseResultConsumer resultConsumer,
+        final SearchPhaseResults<SearchPhaseResult> resultConsumer,
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:46]: `QueryPhaseResultConsumer` is very cumbersome and creates all sorts of things we don't want (pre allocated atomic arrays for one thing). All the behavior is the same among the child classes.

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java (hunk 2 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:571]: I can do that :D

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:799]: Ah yes, I misread!

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/common/util/Maps.java [hunk 5]
```diff
@@ -281,6 +282,18 @@ public static <K, V> Map<K, V> newHashMapWithExpectedSize(int expectedSize) {
...
+    /**
+     * Returns a concurrent hash map with a capacity sufficient to keep expectedSize elements without being resized.
+     *
+     * @param expectedSize the expected amount of elements in the map
+     * @param <K> the key type
+     * @param <V> the value type
+     * @return a new pre-sized {@link HashMap}
+     */
+    public static <K, V> Map<K, V> newConcurrentHashMapWithExpectedSize(int expectedSize) {
+        return new ConcurrentHashMap<>(capacity(expectedSize));
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/common/util/Maps.java:294]: I believe this is unnecessary. Unlike HashMap, ConcurrentHashMap does this calculation in its constructor `long size = (long)(1.0 + (long)initialCapacity / loadFactor);`

**Pull request merged**